diff --git a/cryptroot-unlock b/cryptroot-unlock
index d2b6f51..6105bdf 100755
--- a/cryptroot-unlock
+++ b/cryptroot-unlock
@@ -36,8 +36,14 @@ fi
 # Print the list of PIDs the executed command of which is $exe.
 pgrep_exe() {
 	local exe="$1" pid
-	ps -eo pid= | while read pid; do
-		[ "$(readlink -f /proc/$pid/exe)" != "$exe" ] || printf '%i\n' "$pid"
+	ps | awk '{print $1, $5}' | while read LINE; do
+		set $LINE
+		local pid=$1
+		local cmd=$2
+		if [ "$cmd" == "$exe" ]; then
+			echo $pid
+			break
+		fi
 	done
 }
 
@@ -104,7 +110,7 @@ wait_for_prompt() {
 			break
 		fi
 
-		usleep 100000
+		sleep 0.1
 		timer=$(( $timer - 1 ))
 		if [ $timer -le 0 ]; then
 			echo "Error: Timeout reached while waiting for askpass." >&2
@@ -113,14 +119,14 @@ wait_for_prompt() {
 	done
 
 	# find the cryptsetup process with same $CRYPTTAB_NAME
-	eval $(grep -Ez '^CRYPTTAB_(NAME|TRIED|SOURCE)=' "/proc/$pid/environ" | tr '\0' '\n')
+	eval $(tr '\0' '\n' < "/proc/$pid/environ" | grep -E '^CRYPTTAB_(NAME|TRIED|SOURCE)=')
 	if ! is_device_known "$CRYPTTAB_NAME"; then
 		echo "Error: Refusing to process unknown device $CRYPTTAB_NAME" >&2
 		exit 1
 	fi
 
 	for pid in $(pgrep_exe '/sbin/cryptsetup'); do
-		if grep -Fxqz "CRYPTTAB_NAME=$CRYPTTAB_NAME" "/proc/$pid/environ"; then
+		if tr '\0' '\n' < "/proc/$pid/environ" | grep -Fxq "CRYPTTAB_NAME=$CRYPTTAB_NAME"; then
 			PID=$pid
 			BIRTH=$(stat -c'%Z' "/proc/$PID")
 			return 0;
@@ -139,7 +145,7 @@ wait_for_prompt() {
 wait_for_answer() {
 	local timer=$(( 10 * $TIMEOUT )) dev
 	until [ ! -d "/proc/$PID" ] || [ $(stat -c'%Z' "/proc/$PID") -gt $BIRTH ]; do
-		usleep 100000
+		sleep 0.1
 		timer=$(( $timer - 1 ))
 		if [ $timer -le 0 ]; then
 			echo "Error: Timeout reached while waiting for PID $PID." >&2
